### DISCUSSION TITLE
How to implement global exception handling using ExceptionMapper<T>?

### DISCUSSION THREAD

**Discussion by jehrenzweig-leagueapps** [quarkusio/quarkus#26033] [Q&A] [answered]
I've created a Quarkus demo application using Kotlin, in which I'm trying to globally handle a custom exception when a RESTEasy reactive endpoint throws it.  However, my global exception handler isn't working as written -- it never gets called:
```
import com.example.exceptions.CustomException
import com.example.models.ProblemDetails
import javax.ws.rs.core.Response
import javax.ws.rs.ext.ExceptionMapper
import javax.ws.rs.ext.Provider


@Provider
class CustomExceptionHandler : ExceptionMapper<CustomException> {
    override fun toResponse(e: CustomException): Response {
        var problemDetails = ProblemDetails(
            status = Response.Status.BAD_REQUEST,
            detail = "Something bad happened.",
            data = mapOf("someValue" to e.someValue)
        )

        var response = Response
            .status(problemDetails.status)
            .entity(problemDetails)
            .build()

        return response
    }
}
```

My code is available here:
- https://github.com/jehrenzweig-leagueapps/quarkus-global-exception-handling/

I'm brand new to Quarkus and pretty new to Kotlin, so if I'm getting tripped up on something basic, my apologies in advance.  Any help & pointers are greatly appreciated -- thanks in advance!

[a comment by an automated account is not shown]

**Reply by mschorsch** [reply]
You don't use Resteasy nor Resteasy Reactive. Your code uses Reactive Routes, thats the reason why your ExceptionMapper isn't called.

In Resteasy Reactive you should use `@ServerExceptionMapper` (https://quarkus.io/guides/resteasy-reactive#exception-mapping), but using the JaxRs way (ExceptionMapper) ist also supported.

**Reply by ghost** [reply]
Thank you for pointing out that I'm using Reactive Routes.  And just to make sure I understand: my `build.gradle.kts` file references `quarkus-resteasy-reactive-jackson`, but the syntax I'm using in my controller is defined by `quarkus-reactive-routes`, right?

Also is it _required_ to use `@ServerExceptionMapper` in each of the classes where a given exception can occur?  Or can I declare it once -- in its own class, for example -- and then have that exception handled globally & consistently for all endpoints?  Or is that what the JAX-RS `ExceptionMapper`-based approach should be used for?  Just trying to identify what my options are... I'd rather not have to copy, paste, & maintain the same code block into multiple classes, if at all possible.

**Reply by mschorsch** [reply]
Yes, you reference Reactive Routes and Resteasy Reactive but only use Reactive Routes. You should write your controller in resteasy Reactive syntax. The reference to Reactive Routes can be removed. 

You don't have to use @ServerExceptionMapper. ExceptionMapper ist handled the same as in "normal" JAX-RS. Resteasy Reactive is a JAX-RS compliant Implementation. As far as i can see your ExceptionMapper seems correct.

https://quarkus.io/guides/resteasy-reactive#exception-mapping:

> exception mappers defined in REST endpoint classes will only be called if the exception is thrown in the same class. If you want to define global exception mappers, simply define them outside a REST endpoint class

> You can also declare exception mappers in the JAX-RS way.

**Comment by jehrenzweig-leagueapps** [comment, marked as answer]
I was finally able to get global exception handling the way I'd wanted it, by removing my reference to `io.quarkus:quarkus-reactive-routes` and rewriting my `HelloWorldController` class to use JAX-RS syntax (i.e. `@GET Path(...) @Produces(...)`) instead of Reactive Routes syntax (`@Route(...)`).  Thank you @mschorsch for putting me on the right path!
